Add ray tracing pipeline library support

archimedus opened the pull request:
Extend ray tracing support in Amber with Pipeline Library support.
Also add:
 * include support as well
 * null shader groups support
 * geometry flags support

dj2 commented:
From the description, it sounds like there are 4 separate PRs in here? Can we split them apart so they're easier to review?

What is the reason behind the `INCLUDE` tag? We've historically been hesitant to add an actual include facility, adding the `VIRTUAL_FILE` support to allow `include` tags inside shaders for HLSL, because we've been concerned that once added, Amber files are no longer a single standalone script. It now has support files, it has paths that have to match when sending the files to someone else, etc. (https://github.com/google/amber/issues/222 is a request to allow external shader files)

archimedus commented:
Well, this feature set is just what I needed to have to implement https://gerrit.khronos.org/c/vk-gl-cts/+/14671.

> From the description, it sounds like there are 4 separate PRs in here? Can we split them apart so they're easier to review?

I don't think it is worthy to split it this way: two other points are rather small improvements not worthy for separate issue.

I will move out INCLUDE as this really have few relations to name of this issue, though it will make more pain implementing issue I mentioned, due to ray scene declaration is a repetetive task that INCLUDE fits best.

> It now has support files, it has paths that have to match when sending the files to someone else, etc. (https://github.com/google/amber/issues/222 is a request to allow external shader files)

That was exactly what I thought starting INCLUDE: well, does not it looks like it is time for INCLUDE?

dj2 commented:
Scanning through the gitlab PR, the `INCLUDE` pulls out a few lines, but also makes it harder to read the individual scripts, so from someone just reading the `.amber` files it makes them harder to understand.

For the others, I think geometry flags could be separated, the overall feature is small, but there is a bunch of stuff in there including amberscript parser changes which, would be a lot easier to review if separated from the pipeline library amberscript changes.

I'm not really sure what the null shader group entails from scanning the PR, so it maybe fine to leave as part of this PR.

archimedus commented:
Ok, here we have only library support.

I use `shared_ptr<ShaderGroup>`, due to it is logically correct: `ShaderGroup` really belong to several pipelines.

Please give it a run.

dj2 commented:
This is the first part of the review for what I've gotten through so far. 

Is there a document on pipeline libraries I can read? I'm a bit confused by some of the things being done and am guessing it relates back to how vulkan sets things up.

I'm a bit confused as to why we jam all the things together early, like, why not keep shader groups in their individual Pipeline objects until we get to the vulkan backend and then we can stuff things together as needed? Does that remove the need for the non lib counts of shaders in pipelines?

I guess, an over arching design of how these pieces all fit together would be useful.

archimedus commented:
> Is there a document on pipeline libraries I can read?

Well, I was using vulkan specification only, I cannot suggest anything better for ray tracing pipeline libraries. The most essential part that I found in spec I added as comment. As far as I understand pipeline libraries is mostly separated from graphics pipeline libraries (that is a located in different extension).

> why not keep shader groups in their individual Pipeline objects until we get to the vulkan backend

I think the comment I added is best explains this:

    // Here is a tricky step: add shaders and shader groups from library to
    // pipeline itself. Like spec "suggests":
    //
    // If pipeline libraries are included in  pLibraryInfo, shaders defined in
    // those libraries are treated as if they were defined as additional entries
    // in pStages, appended in the order they appear in the pLibraries array and
    // in the pStages array when those libraries were defined.
    //
    // When referencing shader groups in order to obtain a shader group handle,
    // groups defined in those libraries are treated as if they were defined as
    // additional entries in pGroups, appended in the order they appear in the
    // pLibraries array and in the pGroups array when those libraries were
    // defined.The shaders these groups reference are set when the pipeline
    // library is created, referencing those specified in the pipeline library,
    // not in the pipeline that includes it.

On practical stage: amber file in pipeline (that uses some pipeline library) can refer to shader group names inherited from pipeline libraries used, thus we need to either parse whole tree of used pipeline libraries or do as I just did: include pipeline groups into pipeline that uses pipeline library. Same with shaders (pStages).

> Does that remove the need for the non lib counts of shaders in pipelines?
I am not sure what do you mean here.

As an example I suggest to take a look example I have written for CTS:

https://gerrit.khronos.org/c/vk-gl-cts/+/14671/10/external/vulkancts/data/vulkan/amber/ray_tracing/basic_lib.amber

archimedus commented:
Could you please give another kokoro run?

archimedus commented:
Thank you.

Please another run to make sure that all is fixed.

archimedus commented:
Fixed few more things. I hope this time will go smooth.

archimedus commented:
It seems it's done.

dj2 reviewed (commented):
I just realized, my comments were 'pending'. Apologies for not sending these correctly earlier.

archimedus commented on docs/amber_script.md:
```diff
@@ -589,6 +597,58 @@ Generally a program needs three shader binding tables:
+Specification requires that all pipeline libraries used within pipeline and pipeline using
+libraries itself have same value.
+
+List of used libraries must preceed shader group (SHADER_GROUP) or shader binding tables
```
At parsing time I check whether we have declared shader groups and shader binding tables declared in already parsed libraries. That's why order is matter.

dj2 replied:
Hm, I can't remember if Amberscript has parse order requirements already. If not, I'd prefer not to add one, we should validate after the fact.

archimedus commented on src/amberscript/parser.cc:
```diff
@@ -2037,9 +2080,17 @@ Result Parser::ParsePipelineShaderGroup(Pipeline* pipeline) {
-    if (!r.IsSuccess())
-      return r;
+    bool add_shader_to_pipeline = true;
+    if (!pipeline->GetPipelineLibraries().empty() &&
```
I doubt. Shader group implicitly add a shader to pipeline. No reason to add shader if it present in pipeline library. Though subject is controversial. May be we should declare and add shaders to pipeline explicitly? Though in this case how to differ shader_a in inherited pipeline and shader_a added in this pipeline?

dj2 replied:
I'm not sure I understand your comment. You doubt this condition can happen? Or, that we should check for it?

archimedus replied:
```
PIPELINE raytracing my_rtlib
...
  FLAGS LIBRARY
  SHADER_GROUP g_miss2 miss2
END

PIPELINE raytracing my_rtpipeline
  USE_LIBRARY my_rtlib
  SHADER_GROUP g_miss2a miss2
  SHADER_BINDING_TABLE sbt_miss
    g_miss2
    g_miss2a
  END
END
```

Here we have a dilemma for g_miss2a which shader miss2 to use:
1. Reuse one from my_rtlib
2. Create a new one miss2 shader

Technically p2 is possible and may be even have some sense some day, but for now I follow p1.

archimedus replied:
Done

archimedus commented on src/amberscript/parser.cc:
```diff
@@ -2733,9 +2781,9 @@ Result Parser::ParseRun() {
 
       if (tok == "RAYGEN") {
-        if (!cmd->GetRayGenSBTName().empty())
+        if (!cmd->GetRGenSBTName().empty())
```
Fixed size 4-letter abbrevs for shaders looks enough for me.

dj2 replied:
It's a lot harder to read, the` RayGen` is a lot more explict and easier for folks to understand. Can we change it back to `RayGen`?

archimedus commented on src/amberscript/parser.cc:
```diff
@@ -4352,15 +4402,141 @@ Result Parser::ParseSBT(Pipeline* pipeline) {
-    sbtrecord->SetUsedShaderGroupName(tok, pipeline->GetShaderGroupIndex(tok));
+    sbtrecord->SetUsedShaderGroupName(tok);
     sbtrecord->SetCount(1);
+    sbtrecord->SetIndex(-1);
```
Yes, it is, though I would like to enforce it here.

archimedus commented on src/pipeline.h:
```diff
@@ -457,14 +482,40 @@ class Pipeline {
+  }
+  /// Returns number of shader groups belonging to this pipeline without
+  /// libraries addition.
+  size_t GetNonLibShaderGroupCount() {
```
As I have written above:

> On practical stage: amber file in pipeline (that uses some pipeline library) can refer to shader group names inherited from pipeline libraries used, thus we need to **either parse whole tree of used pipeline libraries or do as I just did**: include pipeline groups into pipeline that uses pipeline library. Same with shaders (pStages).

But to create pipeline I need to know how many stages and groups are this pipeline native.

dj2 replied:
It seems like it would be similer to keep the tree of pipelines and assemble what we need when we create the pipeline. Keep the create script closer to the actual amber script instead of attempting to store some compiled pipeline information.

archimedus commented on src/amberscript/parser.cc:
```diff
@@ -661,6 +671,39 @@ Result Parser::ParsePipelineBody(const std::string& cmd_name,
+    // pLibraries array and in the pGroups array when those libraries were
+    // defined.The shaders these groups reference are set when the pipeline
+    // library is created, referencing those specified in the pipeline library,
+    // not in the pipeline that includes it.
```
I would keep it here: parser is first place where future developer meet this.

dj2 replied:
Sorry, I was referring to the entire block, not just this comment. The combining of libraries into pipelines shouldn't be the responsibility of the parser but a later step.

archimedus replied:
I was thinking that this hint from spec writers is a good point of use to speed-up, but if you prefer slow way, ok I will reimplement.

archimedus replied:
Removed.

archimedus commented:
Please give another try.

archimedus commented:
It seems no more fixes required here.
